pkg/opensearch/opensearch.go

```diff
@@ -45,8 +46,124 @@ func NewOpenSearchDatasource(ctx context.Context, settings backend.DataSourceIns
 func (ds *OpenSearchDatasource) CheckHealth(ctx context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
 	res := &backend.CheckHealthResult{}
 
+	jsonDataStr := req.PluginContext.DataSourceInstanceSettings.JSONData
+	jsonData, err := simplejson.NewJson([]byte(jsonDataStr))
+	if err != nil {
+		res.Status = backend.HealthStatusError
+		res.Message = "Failed to parse settings"
+		return res, nil
+	}
+
+	flavor := jsonData.Get("flavor").MustString("")
+	if flavor != string(client.OpenSearch) && flavor != string(client.Elasticsearch) {
+		res.Status = backend.HealthStatusError
+		res.Message = "No version set"
+		return res, nil
+	}
+
+	_, err = client.ExtractVersion(jsonData.Get("version"))
+	if err != nil {
+		res.Status = backend.HealthStatusError
+		res.Message = "No version set"
+		return res, nil
+	}
+
+	timeField, err := jsonData.Get("timeField").String()
+	if err != nil {
+		res.Status = backend.HealthStatusError
+		res.Message = fmt.Sprintf("time field name is required, err=%v", err)
+		return res, nil
+	}
+
+	db := jsonData.Get("database").MustString()
+	indexInterval := jsonData.Get("interval").MustString()
+	ip, err := client.NewIndexPattern(indexInterval, db)
+	if err != nil {
+		res.Status = backend.HealthStatusError
+		res.Message = fmt.Sprintf("Failed to get index: %s", err)
+		return res, nil
+	}
+	// Same as the frontend check, we generate the time pattern indices for the last six hours
+	indices, err := ip.GetIndices(&backend.TimeRange{From: time.Now().Add(-6 * time.Hour), To: time.Now()})
+	if err != nil || len(indices) == 0 {
+		res.Status = backend.HealthStatusError
+		res.Message = fmt.Sprintf("Failed to get indices: %s", err)
+		return res, nil
+	}
+
+	var index string
+	var body []byte
+	// We try the indices until one successfully queries
+	for _, indexName := range indices {
+		index = indexName
+		osUrl, err := createOpensearchURL(index+"/_mapping/field/"+url.PathEscape(timeField), req.PluginContext.DataSourceInstanceSettings.URL)
+		if err != nil {
+			res.Status = backend.HealthStatusError
+			res.Message = err.Error()
+			continue
+		}
+
+		request, err := http.NewRequestWithContext(ctx, http.MethodGet, osUrl, bytes.NewBuffer(nil))
+		if err != nil {
+			res.Status = backend.HealthStatusError
+			res.Message = err.Error()
+			continue
+		}
+		request.Header = req.GetHTTPHeaders()
+
+		response, err := ds.HttpClient.Do(request)
+		if err != nil {
+			res.Status = backend.HealthStatusError
+			res.Message = err.Error()
+			continue
+		}
+
+		body, err = io.ReadAll(response.Body)
+		if err != nil {
+			res.Status = backend.HealthStatusError
+			res.Message = err.Error()
+			continue
+		}
+
+		if response.StatusCode == 200 {
+			break
+		}
```

Comment:
If I understand what you're aiming at here I think we need to add
`res.Status = backend.HealthStatusUnknown` (or `HealthStatusOk`) here before the `break`. Otherwise if the first index we try fails, the check below on line 134 will always be true and we'll return an error, even if a later index succeeds.